packages/opentelemetry-plugin-user-interaction/src/userInteraction.ts

```diff
@@ -0,0 +1,457 @@
+  }
+
+  /**
+   * Creates a new span
+   * @param element
+   * @param eventName
+   */
+  private _createSpan(
+    element: HTMLElement,
+    eventName: string
+  ): types.Span | undefined {
+    if (!element.getAttribute) {
+      return undefined;
+    }
+    if (element.hasAttribute('disabled')) {
+      return undefined;
+    }
+    const xpath = getElementXPath(element, true);
+    try {
+      const span = this._tracer.startSpan(`${EVENT_CLICK_NAME} ${xpath}`, {
+        attributes: {
+          [AttributeNames.COMPONENT]: this.component,
+          [AttributeNames.EVENT_TYPE]: eventName,
+          [AttributeNames.TARGET_ELEMENT]: element.tagName,
+          [AttributeNames.TARGET_XPATH]: xpath,
+          [AttributeNames.HTTP_URL]: window.location.href,
+          [AttributeNames.HTTP_USER_AGENT]: navigator.userAgent,
+        },
+        parent: this._tracer.getCurrentSpan(),
+        links: [],
+      });
+
+      this._uiMem.set(span, {
+        taskCount: 0,
+      });
+
+      return span;
+    } catch (e) {
+      this._logger.error(this.component, e);
+    }
+    return undefined;
+  }
+
+  /**
+   * Decrement number of tasks that left in zone,
+   *     This is needed to be able to end span when no more tasks left
+   * @param span
+   */
+  private _decrementTask(span: types.Span) {
+    const uiMem = this._uiMem.get(span);
+    if (uiMem) {
+      uiMem.taskCount--;
+      if (uiMem.taskCount === 0) {
+        let endTime;
+        if (uiMem.hrTimeLastTimeout) {
+          endTime = uiMem.hrTimeLastTimeout;
+        }
+        this._tryToEndSpan(span, endTime);
+      }
+    }
+  }
+
+  /**
+   * It gets the element that has been clicked when zone tries to run a new task
+   * @param task
+   */
+  private _getClickedElement(task: AsyncTask): HTMLElement | undefined {
+    if (task.eventName === 'click') {
+      return task.target;
+    }
+    return undefined;
+  }
+
+  /**
+   * Increment number of tasks that are run within the same zone.
+   *     This is needed to be able to end span when no more tasks left
+   * @param span
+   */
+  private _incrementTask(span: types.Span) {
+    const uiMem = this._uiMem.get(span);
+    if (uiMem) {
+      uiMem.taskCount++;
+    }
+  }
+
+  /**
+   * This patches the addEventListener of HTMLElement to be able to
+   * auto instrument the click events
+   * This is done when zone is not available
+   */
+  private _patchElement() {
+    const plugin = this;
+    return (original: Function) => {
+      return function addEventListenerPatched(
+        this: HTMLElement,
+        type: any,
+        listener: any,
+        useCapture: any
+      ) {
+        const patchedListener = (...args: any[]) => {
+          const target = this;
+          const span = plugin._createSpan(target, 'click');
+          if (span) {
+            return plugin._tracer.withSpan(span, () => {
+              const result = listener.apply(target, args);
+              // no zone so end span immediately
```

Comment:
How much value do you think it provides to end the click event span immediately if there is no Zone present?

An alternative when Zone.js is not present is to use [MutationObserver](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver) to determine when the DOM is done changing. 

That approach is used by the Boomerang real user monitoring plugin, see [auto-xhr.js](https://github.com/akamai/boomerang/blob/63ef0a5af20288cc2488d10e605c4b165eec9fb8/plugins/auto-xhr.js#L21). I think it's worth giving their algorithm a read to understand how they identify interesting and non-interesting events and try to determine when an interaction is done without use of Zone.js.

Reply:
> How much value do you think it provides to end the click event span immediately if there is no Zone present?
> 
you see at least that someone clicked the button

> An alternative when Zone.js is not present is to use [MutationObserver](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver) to determine when the DOM is done changing.
> 
> That approach is used by the Boomerang real user monitoring plugin, see [auto-xhr.js](https://github.com/akamai/boomerang/blob/63ef0a5af20288cc2488d10e605c4b165eec9fb8/plugins/auto-xhr.js#L21). I think it's worth giving their algorithm a read to understand how they identify interesting and non-interesting events and try to determine when an interaction is done without use of Zone.js.

This is something which might be done as a new task in new PR, not sure yet if it should be extra functionality to existing plugin, a new plugin, maybe maybe also updating scope managers, or mix of these, but definitely not part of this yet.

Reply:
Yeah, I think this is fine. Once this is in we can keep iterating on it and improving it.